### PR TITLE
Feature/liveness cx 2093

### PR DESCRIPTION
**Pull request by stefaniacardenas**
# Problem
User should be able to record a video and reproduce it

# Solution
By adding `?liveness=true`, once on the face step, a user can record a video and reproduce it.
Liveness feature is currently behind a feature flag.

## Checklist
_put `n/a` if item is not relevant to PR changes_

- [n/a] Have the CHANGELOG, README, MIGRATION, RELEASE_GUIDELINES docs been updated?
- [n/a] Have new automated tests been implemented?
- [x] Have new manual tests been written down?
- [x] Have tests passed locally?
- [x] Have any new strings been translated?

### CODE REVIEW

File: src/components/Confirm/style.css [hunk 1]
```diff
@@ -9,6 +9,11 @@
   justify-content: space-between;
 }
 
+.livenessVideo {
```
Comment (stefaniacardenas) [inline src/components/Confirm/style.css]: This is to ensure responsiveness of the video tag in the confirm screen.

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-350-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-350-pr-onfido-sdk-ui-onfido.surge.sh

File: src/components/Camera/Photo.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+import classNames from 'classnames'
+import style from './style.css'
+
+import { CameraPure, CaptureActions } from './index.js'
+
+export default class Photo extends React.Component<CameraType> {
+  webcam = null
+  interval: ?Visibility
+
+  capture = {
```
Comment (rfreitas) [inline src/components/Camera/Photo.js]: not needed

File: src/components/Camera/Photo.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    start: () => {
+      this.capture.stop()
+      this.interval = Visibility.every(1000, this.screenshot)
+    },
+    stop: () => Visibility.stop(this.interval),
+    once: () => this.screenshot()
+  }
+
+  screenshot = () => {
+    const { onScreenshot } = this.props
```
Comment (rfreitas) [inline src/components/Camera/Photo.js]: this can call a camera util function to remove redundancy with Photo

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-350-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://350-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-350-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-350-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://350-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-350-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-350-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://350-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-350-pr-onfido-sdk-ui-onfido.surge.sh

File: src/components/Camera/Video.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  }
+
+  buttonText = () => {
+    const { i18n } = this.props
+    if (i18n) {
+      return this.state.recording ? i18n.t('capture.liveness.stop') : i18n.t('capture.liveness.start')
+    }
+  }
+
+  livenessButton = () => classNames({[style.stopRecording]: this.state.recording, [style.startRecording]: !this.state.recording })
```
Comment (rfreitas) [inline src/components/Camera/Video.js]: Minor: maybe `buttonClassName` ? It should also be consistent with `buttonText` hence why dropped the `liveness`

File: src/components/Camera/index.js [hunk 5]
```diff
@@ -77,111 +61,62 @@ class CameraPure extends React.Component<CameraPureType> {
   }
 
   render() {
-    const {method, autoCapture, title, subTitle, onUploadFallback, onFallbackClick,
-      onUserMedia, faceCaptureClick, webcamRef, isFullScreen, onWebcamError, i18n} = this.props;
+    const {method, title, subTitle, onUploadFallback, onFallbackClick,
+      onUserMedia, webcamRef, isFullScreen, onWebcamError, i18n, liveness} = this.props;
```
Comment (rfreitas) [inline src/components/Camera/index.js]: I would replace `liveness` with `videoRecording`

File: src/components/Capture/capture.js [hunk 6]
```diff
@@ -59,6 +59,12 @@ class Capture extends Component {
     actions.createCapture({method, capture, maxCaptures: this.maxAutomaticCaptures})
   }
 
+  createLivenessVideo(liveness, url) {
+    const payload = {liveness, url}
```
Comment (rfreitas) [inline src/components/Capture/capture.js]: what does `liveness` mean here?

File: src/components/Capture/capture.js [hunk 7]
```diff
@@ -85,6 +91,11 @@ class Capture extends Component {
     })
   }
 
+  onVideoRecorded = (blob) => {
+    const url = window.URL.createObjectURL(blob);
```
Comment (rfreitas) [inline src/components/Capture/capture.js]: `createObjectURL` needs a `revokeObjectURL()` when it's no longer needed. This can be added in a separate ticket, but we cannot forget it.

Comment (stefaniacardenas) [reply src/components/Capture/capture.js]: Yes! I was thinking about that late on Friday!

File: src/components/Capture/capture.js (hunk 6 above)
Comment (stefaniacardenas) [reply src/components/Capture/capture.js]: It's just a parameter I pass to the capture, so that the `Confirm` screen knows it has to render the video tag instead of the `CaptureViewer` component.

**Comment by rfreitas** [comment]
Travis automatic deployment: https://350-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-350-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-350-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://350-pr-onfido-sdk-ui-onfido.surge.sh

File: src/components/Camera/index.js [hunk 8]
```diff
@@ -77,111 +61,62 @@ class CameraPure extends React.Component<CameraPureType> {
...
-      this.capture.stop()
-      this.interval = Visibility.every(1000, this.screenshot)
-    },
-    stop: () => Visibility.stop(this.interval),
-    once: () => {
-      this.screenshot()
-    }
-  }
-
   constructor (props: CameraType) {
```
Comment (TheComfyChair) [inline src/components/Camera/index.js]: minor nitpick: If you're not doing anything except calling super(props), this constructor can be dropped :)

File: src/components/Camera/AutoCapture.js [hunk 9]
```diff
@@ -0,0 +1,50 @@
+// @flow
+import * as React from 'react'
+import { h } from 'preact'
+import Visibility from 'visibilityjs'
```
Comment (TheComfyChair) [inline src/components/Camera/AutoCapture.js]: This library is nice to know about, as it could prove useful for tracking user actions. How often a user needs to tab away from CP would be a great metric for seeing how successful developments like docupedia have been at reducing overall 'non-CP' time a user has to be spent in order to complete a document task.

File: src/components/Camera/Photo.js [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+import style from './style.css'
+import { screenshot } from '../utils/camera.js'
+
+import { CameraPure, CaptureActions } from './index.js'
+
+export default class Photo extends React.Component<CameraType> {
+  webcam = null
+
+  screenshot = () => screenshot(this.webcam, this.props.onScreenshot)
+  buttonText = () => {if (this.props.i18n) return this.props.i18n.t('capture.face.button')}
```
Comment (TheComfyChair) [inline src/components/Camera/Photo.js]: Does this mean we don't have a fallback available when i18n isn't passed as a prop? It's not something that should happen (this particular component isn't exposed directly to any external developers I think?), so I don't think there's a change required. It's just to make sure I'm not missing anything.

File: src/components/Camera/Video.js [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+type VideoState = {
+  recording: boolean
+}
+
+export default class Video extends React.Component<CameraType, VideoState> {
+  webcam = null
+
+  constructor(props: CameraType) {
+    super(props)
+    this.state = { recording: false }
```
Comment (TheComfyChair) [inline src/components/Camera/Video.js]: A little later on the state is defined using the class property syntax. Both are totally valid (I prefer the class variable syntax nowadays, as there's less to write) but it's probably worth making them consistent :)

Comment (rsilva4) [reply src/components/Camera/Video.js]: > A little later on the state is defined using the class property syntax.

Sorry didn't understand the phrase and the context confused even more, can you rephrase please?

File: src/components/Camera/AutoCapture.js (hunk 9 above)
Comment (rsilva4) [reply src/components/Camera/AutoCapture.js]: Maybe this comment could be share on an internal forum instead since it is a bit out of scope for this public PR. ;)

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-350-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://350-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-350-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-350-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://350-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-350-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-350-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://350-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-350-pr-onfido-sdk-ui-onfido.surge.sh
